src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1389,6 +1464,87 @@ else if (path.toFile().getName().contains("sun")) {
+                if (lastRuleNumberPartsAmount <= partIndex) {
+                    // equal up to here and last rule has less parts,
+                    // thus order is correct, stop comparing
+                    break;
+                }
+
+                final String ruleNumberPart = ruleNameParts[partIndex];
+                final String lastRuleNumberPart = lastRuleNameParts[partIndex];
+                final boolean ruleNumberPartsAreNumeric = IntStream.concat(
+                        ruleNumberPart.chars(),
+                        lastRuleNumberPart.chars()
+                ).allMatch(Character::isDigit);
+
+                if (ruleNumberPartsAreNumeric) {
+                    final int numericRuleNumberPart = parseInt(ruleNumberPart);
+                    final int numericLastRuleNumberPart = parseInt(lastRuleNumberPart);
+                    Assert.assertThat(
+                            outOfOrderReason,
+                            numericRuleNumberPart < numericLastRuleNumberPart,
+                            describedAs("'%0' should not be less than '%1'",
+                                    is(false),
+                                    numericRuleNumberPart, numericLastRuleNumberPart));
+                }
+                else {
+                    Assert.assertThat(
+                            outOfOrderReason,
+                            ruleNumberPart.compareToIgnoreCase(lastRuleNumberPart) > 0,
+                            describedAs("'%0' should not be greater than '%1'",
+                                    is(false),
+                                    ruleNumberPart, lastRuleNumberPart));
+                }
+                lastRuleNumberPartWasEqual = ruleNumberPart.equalsIgnoreCase(lastRuleNumberPart);
+            }
+            if (ruleNumberPartsAmount == partIndex && lastRuleNumberPartWasEqual) {
+                if (lastRuleNumberPartsAmount == partIndex) {
+                    final String ruleNamePart = ruleNameParts[partIndex];
+                    final String lastRuleNamePart = lastRuleNameParts[partIndex];
+                    final int order = ruleNamePart.compareToIgnoreCase(lastRuleNamePart);
+                    Assert.assertThat(
+                            fileName + " rule '" + ruleName + "' is appearing twice",
+                            order,
+                            describedAs("'%0' should not be equal to '%1'",
+                                    is(not(0)),
+                                    ruleName, lastRuleName));
+                    Assert.assertThat(
+                            outOfOrderReason,
+                            order > 0,
+                            describedAs("'%0' should be greater than '%1'",
+                                    is(true),
+                                    ruleNamePart, lastRuleNamePart));
+                }
+                else {
+                    Assert.fail(outOfOrderReason);
```

Comment:
When would this be hit that the other section wouldn't catch?

Reply:
This is the case where the last rule number had more parts than the current, but all were equal so far, like:
```
1.2.3 foo
1.2 bar
```